hydropandas/io/water_connect.py

```diff
@@ -0,0 +1,388 @@
+import json
+import logging
+import zipfile
+from io import BytesIO, StringIO
+from pathlib import Path
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import requests
+from platformdirs import user_data_dir
+from shapely.geometry import box
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+DH_zip_url = (
+    "https://www.waterconnect.sa.gov.au/Content/Downloads/DEW/WATER_Drillholes_shp.zip"
+)
+
+
+def get_obs_list_from_extent(
+    extent,
+    ObsClass,
+    tmin=None,
+    tmax=None,
+    only_metadata=False,
+    keep_all_obs=False,
+    location_gdf=None,
+    **kwargs,
+):
+    """Get observations within a specific extent and optionally for a specific location
+    and grootheid_code.
+
+    Parameters
+    ----------
+    extent : list, tuple, numpy-array or None, optional
+        get water connect measurements within this extent
+        [xmin, xmax, ymin, ymax]
+    ObsClass : type
+        class of the observations, e.g. GroundwaterObs
+    tmin : str or None, optional
+        start time of observations. The default is None.
+    tmax : str or None, optional
+        end time of observations. The default is None.
+    only_metadata : bool, optional
+        if True download only metadata, significantly faster. The default
+        is False.
+    keep_all_obs : bool, optional
+        if False, only observations with measurements are kept. The default
+        is True.
+    location_gdf : GeoDataFrame, optional
+        geodataframe with the locations of the water drill holes you want to include.
+    **kwargs
+        additional keyword arguments are passed to the ObsClass.from_waterinfo()
+        method
+
+    Returns
+    -------
+    obs_list : list
+        list with Obs objects
+
+    """
+
+    if only_metadata and not keep_all_obs:
+        logger.error(
+            "you will get an empty ObsCollection with only_metadata is True and"
```

Comment:
Why? Seems like these two don't have to be exclusive? If i just want metadata why do I need to keep all obs?

Reply:
I get what you mean but I would not want to support it because of the runtimes. The 'only_metadata' option is there to rapidly get all metadata without having to download the observations. If you want to 'keep_all_obs' you have to download all the observations to know if they are there. If you want to have only metadata of the observations with measurement it would be just as fast to set 'keep_all_obs' to True and filter empty observations out later. I improved the logging a bit and moved the level to warning and set 'keep_all_obs' manually to True. So no error will be raised, only the 'keep_all_obs' will be ignored.

Reply:
That sounds like a good solution :+1: